src/components/library/LibraryMangaGrid.tsx

```diff
@@ -48,30 +48,25 @@ const queryGenreFilter = (query: NullAndUndefined<string>, { genre }: IMangaCard
     return queries.every((element) => genre.map((el) => el.toLowerCase()).includes(element));
 };
 
+const performSearch = (searchStr: NullAndUndefined<string>, mangas: IMangaCard[]): IMangaCard[] => {
+    if (!searchStr) {
+        return [...mangas];
+    }
+
+    const titleFilteredManga = mangas.filter((m) => queryFilter(searchStr, m));
+    const genreFilteredManga = mangas.filter((m) => queryGenreFilter(searchStr, m));
+    const uniqueMangas = new Set([...titleFilteredManga, ...genreFilteredManga]);
+
+    return Array.from(uniqueMangas);
+};
+
 const filterManga = (
     manga: IMangaCard[],
     query: NullAndUndefined<string>,
     unread: NullAndUndefined<boolean>,
     downloaded: NullAndUndefined<boolean>,
-): IMangaCard[] => {
-    let filteredManga: IMangaCard[] = [];
-    if (query) {
-        const titleFilteredManga = manga.filter((m) => queryFilter(query, m));
-        const genreFilteredManga = manga.filter((m) => queryGenreFilter(query, m));
-        const unique = titleFilteredManga.concat(genreFilteredManga).reduce((acc: Record<number, IMangaCard>, obj) => {
-            const { id } = obj;
-            if (!acc[id]) {
-                acc[id] = obj;
-            }
-            return acc;
-        }, {});
-        filteredManga = Object.values(unique);
-    }
-    filteredManga = (filteredManga.length ? filteredManga : manga).filter(
-        (m) => downloadedFilter(downloaded, m) && unreadFilter(unread, m),
-    );
-    return filteredManga;
-};
+): IMangaCard[] =>
+    performSearch(query, manga).filter((m) => downloadedFilter(downloaded, m) && unreadFilter(unread, m));
```

Comment:
filtering the manga, before calling performSearch should increase the performance for large libraries

Reply:
you're right it's not optimal currently.
I think the best thing should be to perform search and filter on each manga in one go.
I'll open a new pr for that

Reply:
nervermind, changed it in this pr